### ISSUE TITLE
PortableGtkReactor is broken

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#3203]
|<img alt="jdahlin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| jdahlin reported|
|-|-|
|Trac ID|trac#3203|
|Type|release blocker: regression|
|Created|2008-04-23 20:47:52Z|
|Branch|https://github.com/twisted/twisted/tree/portablegtkreactor-3203-2|

I get the following error when trying to use it on win32:
AttributeError: 'PortableGtkReactor' object has no attribute '_simtag'

The workaround is to set self._simtag = None in the constructor.

Attachments:

* [3203-win.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ad4/ad4577d7a9dce895d5d4991150a1fe7b6d5aa745/929f97b547e429292098020e1ea65886d0decf50.diff) (948 bytes) - added by therve on 2008-05-06 18:39:23Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__3203 3203
type__release_blocker__regression release blocker: regression
reporter__jdahlin jdahlin
priority__normal normal
milestone__ 
branch__branches_portablegtkreactor_3203_2 branches/portablegtkreactor-3203-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1208983672000000 1208983672000000
changetime__1211624527000000 1211624527000000
version__None None
owner__ 

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [23422]) Branching to 'portablegtkreactor-3203'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_3"></a>|@glyph commented|
|-|-|

[#3221](https://github.com/twisted/twisted/issues/3221) was a duplicate of this bug.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun commented|
|-|-|

The full test suite hangs when run against the branch.  It gets to a test in flow and that's the end.  Not sure what's up, and I don't have a Windows machine so debugging further will be more of a hassle for me than I'm likely to be willing to deal with.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_5"></a>|@therve set owner to @therve|
|-|-|

I'll try to look at it.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve commented|
|-|-|

The attached patch fixes most problems I saw. There is still one problem with some tests succeeding after timeout (like LargeBufferTestCase.testWriter). But the rest looks good.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_7"></a>|@therve commented|
|-|-|

(In [23528]) Apply patch

Refs [#3203](https://github.com/twisted/twisted/issues/3203)

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_8"></a>|@therve removed owner|
|-|-|

Please review.

**Comment by twisted-trac** [comment]
|<img alt="jdahlin's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_9"></a>|jdahlin commented|
|-|-|

gtk.mainquit() is a leftover since PyGTK 0.6.x (gtk 1.2.x) and all references to it should probably be removed, since the module is called gtk2reactor after all.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun commented|
|-|-|

(In [23689]) Branching to 'portablegtkreactor-3203-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @therve|
|-|-|

Merged forward to make sure it passed at least as many tests on trunk, and it seems to.  Change seems sensible to me, mostly mirroring what was done to the other gtk2reactor I guess.  I guess we don't have any test coverage showing that `gtk.mainquit()` is doing anything.  I dunno what versions of stuff anyone has or would expect to work.  I'd be inlined to leave that code alone for a while longer, since all you did was indent it.

So, seems good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_12"></a>|@therve set status to `closed`|
|-|-|

(In [23696]) Merge portablegtkreactor-3203-2

Authors: therve, exarkun
Reviewer: exarkun
Fixes [#3203](https://github.com/twisted/twisted/issues/3203)

Fix PortableGtkReactor: we're now able to run trial test suite with the gtk2
reactor under Windows.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
